Add `sum` function in meter system

kezhenxu94 opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👇 ====
### Add an agent plugin to support <framework name>
- [ ] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md)
- [ ] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml)
- [ ] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets)
     ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->

<!-- ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👇 ====
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.
     ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👆 ==== -->

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>. NO
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).


The MAL didn't implement the `sum` function and the E2E test of the previous #6388 actually falls into the `avg` function, causing all `log_count_info` is `1`, this PR adds the real implementation of `sum` function and adjust the E2E test.

<img width="415" alt="image" src="https://user-images.githubusercontent.com/15965696/108861088-d1c4bb80-7629-11eb-808d-1e709f828398.png">

wu-sheng commented on oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java:
```diff
@@ -175,7 +176,10 @@ public SampleFamily min(List<String> by) {
-        ExpressionParsingContext.get().ifPresent(ctx -> ctx.aggregationLabels.addAll(by));
+        ExpressionParsingContext.get().ifPresent(ctx -> {
+            ctx.aggregationLabels.addAll(by);
+            ctx.downsampling = DownsamplingType.AVG;
```
AVG seems default in `ExpressionParsingContext#create`. Is this for making codes more clear or functional need?

kezhenxu94 replied:
> AVG seems default in `ExpressionParsingContext#create`. Is this for making codes more clear or functional need?

`avg` being default had no problem before since it was only one function, now since I added another one, if the expression contains multiple function, and the `avg` being last one, it may be wrong.  For example, say the expression is `log_count.sum(by: ['service', 'instance']).avg(['level'])`, (this expression may make no sense, just for demonstration), the final metrics should be aggregated by `avg` instead of `sum`.

wu-sheng replied:
Got it. A suggestion, could you use this PR to format `SampleFamily` a little better? Such as put `avg` in this way
```
    public SampleFamily avg(List<String> by) {
        ExpressionParsingContext.get().ifPresent(ctx -> ctx.aggregationLabels.addAll(by));
        if (this == EMPTY) {
            return EMPTY;
        }
        if (by == null) {
            double result = Arrays.stream(samples).mapToDouble(s -> s.value).average().orElse(0.0D);
            return SampleFamily.build(this.context, newSample(ImmutableMap.of(), samples[0].timestamp, result));
        }
        return SampleFamily.build(
            this.context,
            Arrays.stream(samples)
                  .map(sample -> 
                           Tuple.of(
                               by.stream()
                                 .collect(toImmutableMap(
                                     labelKey -> labelKey, 
                                     labelKey -> sample.labels.getOrDefault(labelKey, ""))
                                 ), 
                               sample)
                  )
                  .collect(groupingBy(Tuple2::_1, mapping(Tuple2::_2, toList())))
                  .entrySet().stream()
                  .map(entry -> 
                           newSample(entry.getKey(), 
                                     entry.getValue().get(0).timestamp, 
                                     entry.getValue().stream().mapToDouble(s -> s.value).average().orElse(0.0D)))
                  .toArray(Sample[]::new)
        );
    }
```

wu-sheng replied:
Including several other methods. The codes mixed Tuple, it is hard to read.

kezhenxu94 commented on docs/en/concepts-and-designs/mal.md:
```diff
@@ -177,23 +177,19 @@ MAL should instruct meter-system how to do downsampling for metrics. It doesn't
-or
-
-latest last_server_state_sync_time_in_seconds.tagEqual('production', 'catalog')
+last_server_state_sync_time_in_seconds.tagEqual('production', 'catalog').latest()
```
This may be a small breaking change, the form of `latest xxxx` is now NOT supported, instead, one should use `xxx.latest()`. The form of `latest xxx` is error-prone b/c something like

```
latest (foo - 1).tagEqual('bar', '1').sum(['tt']).irate().histogram().histogram_percentile([50,99]).service(['rr'])
```

should be considered the `latest` value of `(foo - 1).tagEqual('bar', '1').sum(['tt']).irate().histogram().histogram_percentile([50,99]).service(['rr'])`, but it's actually the `histogram_percentile` of `latest((foo - 1).tagEqual('bar', '1')).sum(['tt']).irate().histogram()`.

wu-sheng replied:
This looks good to me, @hanahmily any idea?

hanahmily replied:
LGTM. and you have to search our mal relevant rules to adapt to this change.

kezhenxu94 replied:
Done, only unit tests used this

wu-sheng commented:
Test fails, labels missing
>2021-02-24T12:37:06.7762470Z [ERROR]   ExpressionParsingTest.test:124 
2021-02-24T12:37:06.7764030Z Expected: is <ExpressionParsingContext(samples=[foo], isHistogram=true, percentiles=[50, 99], aggregationLabels=[], scopeLabels=[rr], downsampling=LATEST, scopeType=SERVICE)>
2021-02-24T12:37:06.7766583Z      but: was <ExpressionParsingContext(samples=[foo], isHistogram=true, percentiles=[50, 99], aggregationLabels=[tt], scopeLabels=[rr], downsampling=LATEST, scopeType=SERVICE)>

wu-sheng commented:
Duplicated adding? The error log seems missing label. What do I miss?

kezhenxu94 commented:
> Duplicated adding? The error log seems missing label. What do I miss?

It WAS missing, after I added them in the tests, it shows up multiple times because I added multi-aggregation-functions cases (e.g. `xxx.avg(['a']).sum(['a'])`, causing `aggregationLabels` being `['a', 'a']`

wu-sheng commented:
I am just being informed this change will cause a serious bug. The previous `sum`(with avg downsampling) is correct, such as calculating the disk usage, CPU etc. Then, the SampleFamily have 30%(for example), and 35% next time, then we really need 33% usage. `Sum function with avg downsampling` is widely used.

wu-sheng reviewed (approved):
This seems very clear to me. Thanks for polishing this.

wu-sheng commented:
FYI @wankai123 @hanahmily
